Add octave build commands

matsievskiysv opened the pull request:
Add support for build with octave.
Some functions do not work with octave though. More work is needed. Maybe this could be pulled into separate branch?

lfarv commented:
Hello @matsievskiysv,
I have a few comments on your proposals:
- Concerning the build commands: can you try to avoid makefiles ? These are not portable across architectures, and may need some tuning by the user. Using the `mkoctfile --mex`  or `mex` command within Octave should provide a universal way of building extensions, including on Windows. Is there any problem with that (I am not familiar with Octave)? Also, rather than modifying `atmexall` (already complicated enough), you can provide a dedicated build function for Octave.
- About missing Matlab functions (ex.: `endsWith`): one could add a top-level directory (`atoctave` for instance). It should be ignored by Matlab users (only `atmat` and `atintegrators` in the path), and should be put in 1st position by Octave users. There, you could put any function that simulates a missing Matlab function, or that replaces an `atmat` function for Octave users. Then you can keep the name `endsWith` (for instance) wherever it appears and provide your own version in `atoctave`.
- You can forget the files in the `atintegrators/user` directory: they are still there for historical reasons, by have not been upgraded for the python interface and are not compiled!

Finally, I agree with you idea of temporarily creating a dedicated branch for octave, until it can be merged into the main one.

matsievskiysv commented:
@lfarv
> Concerning the build commands: can you try to avoid makefiles ...
  
Replaced with file `bootstrapOctave.m`

> About missing Matlab functions ...
  
Done

> You can forget the files ...

Removed them from source list

> Finally, I agree with you idea of temporarily creating a dedicated branch for octave, until it can be merged into the main one.

As soon as you'll create an `octave` branch, I'll set it as this PR's merge target.

lfarv commented:
There is now a branch `octave_compatibility`. It adds an `atoctave` directory, where you should put all octave-specific .m files, so that they are ignored in Matlab.

matsievskiysv commented:
Ok. Demos seems to work except for the GUI stuff and ones that fail because of #200

matsievskiysv commented:
I don't know how the output of the demos should look like, so I attach graphical output:

---

**atfastring/testatfastring.m**

![fastring-1](https://user-images.githubusercontent.com/26824900/100719496-c74da980-33cd-11eb-82be-f60f42ea78f8.png)
![fastring-2](https://user-images.githubusercontent.com/26824900/100719510-cae13080-33cd-11eb-8f02-3db38b99b576.png)

---

**atmatchExamples/bump/run_BumpFit.m**

![bump](https://user-images.githubusercontent.com/26824900/100719607-e64c3b80-33cd-11eb-81d8-c5473022cee3.png)

---

**atmatchExamples/ExampleATMATCH/runtotest_atmatch.m**

![exampleATMATCH](https://user-images.githubusercontent.com/26824900/100719710-0bd94500-33ce-11eb-8cbc-00ceb6bb76d3.png)

---

**IDModeling/trackWithU35KickMap.m**

![trackWithU35KickMap-1](https://user-images.githubusercontent.com/26824900/100719796-27445000-33ce-11eb-9366-a1db2708bcf7.png)

![trackWithU35KickMap-2](https://user-images.githubusercontent.com/26824900/100719821-30352180-33ce-11eb-98bb-a3e196444734.png)

---

**opticsAndBeamSizes/linoptdemo.m**

*This one is obviously messed up*

![linpotdemo](https://user-images.githubusercontent.com/26824900/100719989-6d99af00-33ce-11eb-8ae6-1f9717fbac94.png)

---

**opticsAndBeamSizes/ohmienvelopedemo.m**

![ohmienvelopedemo-1](https://user-images.githubusercontent.com/26824900/100720092-8d30d780-33ce-11eb-8f0a-3d7ff29834ab.png)

![ohmienvelopedemo-2](https://user-images.githubusercontent.com/26824900/100720103-902bc800-33ce-11eb-9335-2195da1daa59.png)

---

**responseMatrix/findrespmdemo.m**

![findrespmdemo-1](https://user-images.githubusercontent.com/26824900/100720146-a5085b80-33ce-11eb-8c99-81d6d105b853.png)

![findrespmdemo-2](https://user-images.githubusercontent.com/26824900/100720159-a9347900-33ce-11eb-9379-6a03b9695a62.png)

lfarv commented:
Hi @matsievskiysv, That's excellent ! There are still a few problems, but it's already very promising. On the demos, there is a problem with H/V coupling: you pointed the problem on atinopt, bit on ohmienvelope, on the right axis, the scale should be in the 1E-4 while you get 1E-18. Also, in findrespmdemo, the off-diagonal blocks are at zero, which is normal for the non-coupled ring, but not for the coupled ring. Finally, a detail, on some plots (bump fit, attach), you should get a synoptic representation of the ring magnets along the horizontal axis. I'll think about what could cause the tracking problems for could machines: it's like the tilted magnets would not couple horizontal position into the vertical plane…

lfarv commented:
Sorry to intervene again:

- For compilation of integrators: you should select  all ‘*Pass.c’ and ‘*Pass.cc’ files only. All other *.c or *.cc files are meant to be ‘#include’d in integrators, but not compiled independently.
- Can you avoid modifying all unused ‘atintegrators/user/*.c files ? The less modified files, the easier the review…
- I am worried by the requirement for the un-released Octave version 6. In my experience, system administrators are reluctant to continuously upgrade software packages to the last version, and users have often to leave with slightly older packages. This could block many potential users. I don't know how many functions are affected by the bug you mention, but at least the most critical ones could be re-written in the `atoctave` directory by extracting the problematic nested functions. For others, a partially functioning AT could be better than nothing.

lfarv commented:
For info: did you try `atdiag`? Here is the expected output:

![test](https://user-images.githubusercontent.com/20038121/100751394-8c586f80-33e7-11eb-99d2-f98966da9f4e.png)

matsievskiysv commented:
@lfarv  About `function_handle` for nested functions. It seems that I can wrap them into anonymous function:
```octave
[x,y,z] = dostuff(@handle, arg)		% replace this
[x,y,z] = dostuff(@(a) handle(a), arg)	% with this
```
Is this acceptable?

matsievskiysv commented:
By the way, Octave 6.1 [is out](https://www.gnu.org/software/octave/news/release/2020/11/26/octave-6.1.0-released.html)

matsievskiysv commented:
I get this for `atdiag`:
![diag](https://user-images.githubusercontent.com/26824900/100754653-5f628780-33fc-11eb-8315-9591ba8de5b6.png)

Plot legend and annotations are missing.

matsievskiysv commented:
> For compilation of integrators: you should select all ‘*Pass.c’ and ‘*Pass.cc’ files only. All other *.c or *.cc files are meant to be ‘#include’d in integrators, but not compiled independently.

Edited

> Can you avoid modifying all unused ‘atintegrators/user/*.c files ? The less modified files, the easier the review…

Removed edits

> I am worried by the requirement for the un-released Octave version 6. In my experience, system administrators are reluctant to continuously upgrade software packages to the last version, and users have often to leave with slightly older packages. This could block many potential users. I don't know how many functions are affected by the bug you mention, but at least the most critical ones could be re-written in the atoctave directory by extracting the problematic nested functions. For others, a partially functioning AT could be better than nothing.

Unfortunately for now only 6.1 works. There's also some incompatibility in line continuations (`...`).

matsievskiysv commented:
> after small alterations of atplotsyn I get this

It also helped with some other plots

![bump](https://user-images.githubusercontent.com/26824900/100760854-44dfdc80-3403-11eb-8c26-33913686b469.png)
![atmatch1](https://user-images.githubusercontent.com/26824900/100760872-47dacd00-3403-11eb-8055-d86269a27917.png)
![atmatch2](https://user-images.githubusercontent.com/26824900/100760876-490bfa00-3403-11eb-8e00-6fe7dc5a4f81.png)
![atmatch3](https://user-images.githubusercontent.com/26824900/100760879-490bfa00-3403-11eb-92f7-fd84e5a941bb.png)

lfarv commented on atmat/attrack/linepass.m:
```diff
@@ -90,7 +90,7 @@
 newlattice = double(~keeplattice);
 
-[prefunc,postfunc]=parseargs({function_handle.empty,function_handle.empty},...
+[prefunc,postfunc]=parseargs({cell(),cell()},...
```
Works, but isn't there any way to create empty arrays on any type in Octave?

matsievskiysv replied:
There's no `function_handle` class in Octave. In any case, anything empty will work: https://github.com/atcollab/at/blob/master/atmat/attrack/atpass.c#L403-L404

lfarv commented on atmat/pubtools/lattice_tools/atreadbeta.m:
```diff
@@ -160,7 +160,7 @@
 
 
-evalin('base','global GLOBVAL');
+assignin('base','GLOBVAL',GLOBVAL);
```
It is different from the original: here you create a local variable in the base workspace, identical to the global one (bad idea). Declaring the global variable in the base workspace makes it visible, but that's not so useful. The best is to suppress completely these statements (and the similar ones)

matsievskiysv replied:
Original statement made Octave crush. If these are not needed, I will gladly remove them.

lfarv replied:
Go on, remove all the similar lines that you pointed out

matsievskiysv replied:
It broke some code. Particularly, `machine_data`. I'll need some time to fix it

lfarv reviewed (commented):
I do not see incompatibilities with Matlab. One could merge this into the octave branch and see!

lfarv commented:
Now we are left with the problem of wrong results for coupled machine. I cannot get Octave 6.1 on MacOS, and I understood that you don't have access to Matlab. That makes checks difficult… The simpler check is just comparing numerical results for pure tracking over 1 turn of a coupled machine, then go to closed orbit, and then to transfer matrices. If you do not see an easier way, I could send you .mat files with Matlab results, can you read them? It may be slow since I cannot spend so much time on AT, but we can try.

matsievskiysv commented:
> I could send you .mat files with Matlab results, can you read them

I [should be able to](https://stackoverflow.com/questions/9915658/how-do-you-open-mat-files-in-octave), if `.mat` versions will match.

Maybe I'll be able to extract some information from https://github.com/atcollab/ATtraining examples as well.

matsievskiysv commented:
I've created a [docker](https://www.docker.com/) image with a [jupyter server](https://jupyter.org/) with [Octave](https://github.com/calysto/octave_kernel) and AT. Not all plots show, but it's sufficient for testing purposes.

@lfarv If you'll give me your email address, I could send you a link to my personal server so you could test AT there.

![image](https://user-images.githubusercontent.com/26824900/101020386-7978a300-357f-11eb-95e6-cfdc7204c643.png)

![image](https://user-images.githubusercontent.com/26824900/101020407-872e2880-357f-11eb-944c-afc97d7f8764.png)

In order to use it:
1. Install [Docker](https://docs.docker.com/docker-for-mac/install/) (should be trivial)
1. Run command `docker run -d --name at -p 8888:8888 matsievskiysv/at`
1. Run `docker logs at --follow` and write down connection token (will appear on screen as `http://127.0.0.1:8888/?token=<token>`)
1. From web-browser go to `http://127.0.0.1:8888/` page and insert token there.
1. Create a notebook and execute commands. All AT should be available.

willrogers commented:
This could be a really useful development, particularly allowing us to run Matlab tests and Python tests on pull requests and maybe compare the results. I'm very pleased someone is trying this, thanks!

I see Octave 6.1 was released on November 26th, but anyway since we have no Octave support at all at the moment I wouldn't be worried that it requires a new version.

lfarv commented:
@matsievskiysv:

- I'll try the 'docker' way as soon as I'm back home (next monday probably). How shall I get your image?. If that fails, I'll try the access to your server,
- I propose that you continue the development of the octave port directly on the `octave_compatibility` branch. It will remove one intermediate step.

matsievskiysv commented:
> How shall I get your image?

`docker run -d --name at -p 8888:8888 matsievskiysv/at` command will download [it](https://hub.docker.com/r/matsievskiysv/at) automatically.

> I propose that you continue the development of the octave port directly on the octave_compatibility branch. It will remove one intermediate step.

It's fine by me

lfarv commented:
Here is a set of 2 files for checking `atlinop` with octave:

1. Remove the .txt extension which I added to get them accepted as attachments and put them anywhere in the octave path
2. execute checklinopt with either;

- `[results,expected]=checklinopt('coupled');`
- `[results,expected]=checklinopt('uncoupled');`

`results` and `expected` are structures containing the output of `atlinopt` in `lindata`, `tunes`, `chrom` fields.
'coupled' and 'uncoupled' run the same ring (spear2 example) without and with H/V coupling, introduced with quadrupole rotations.

[checklinopt.m.txt](https://github.com/atcollab/at/files/5637601/checklinopt.m.txt)
[checkring.mat.txt](https://github.com/atcollab/at/files/5637602/checkring.mat.txt)

matsievskiysv commented:
@lfarv Results look identical

matsievskiysv commented:
Got jupyter plots working. They do not look as good as Qt plots unfortunately.

![image](https://user-images.githubusercontent.com/26824900/101087539-64256800-35c3-11eb-8b40-654f067dda67.png)

matsievskiysv commented on atmat/atmatch/atmatch.m:
```diff
@@ -131,7 +131,7 @@
 
-options=optimset(minimizer);
+options=optimset(func2str(minimizer));
 
```
Done

lfarv commented:
@matsievskiysv: you should now be allowed to merge you fork into this branch. Let me know

matsievskiysv commented:
Yes, I can merge now. But I want to fix `evalin` issues first

matsievskiysv commented:
Don't know exactly why, but response matrix for coupled lattice now looks like this:

![image](https://user-images.githubusercontent.com/26824900/101146998-d41c0880-362c-11eb-92c7-481d3a603037.png)

lfarv commented:
For `evalin` problems: some demos are using a obsolete way of keeping the 'active' ring structure in the global variable "`THERING`". Though this should still work, it is now recommended to set the ring structure as 1st argument of most functions.

For response matrix: it now looks good!

For `atlinopt`: fine if numerical results are good. I found out that the `linopt` demo take a non coupled ring (spear2), and introduces coupling by adding randomly generated tilts. This is not a good idea since the demo will give different results each time it is run!  But this does not explain why the coupling looked zero in you plot. Did you try again since the response matrix looks good?

matsievskiysv commented:
`linoptdemo`:
![image](https://user-images.githubusercontent.com/26824900/101148576-db441600-362e-11eb-8a52-97e2d156222e.png)
`ohmienvelopedemo`:
![image](https://user-images.githubusercontent.com/26824900/101148669-fdd62f00-362e-11eb-91b0-ae6465291afa.png)
![image](https://user-images.githubusercontent.com/26824900/101148703-0595d380-362f-11eb-9746-353712467cbe.png)

matsievskiysv commented:
Jupyter inline graphics looks a bit sad, but works.

![image](https://user-images.githubusercontent.com/26824900/101149845-86a19a80-3630-11eb-9546-43089bcc6717.png)
